gloo/transport/tcp/socket.cc

```diff
@@ -0,0 +1,175 @@
+  int value = on ? 1 : 0;
+  auto rv = ::setsockopt(fd_, IPPROTO_TCP, TCP_NODELAY, &value, sizeof(value));
+  GLOO_ENFORCE_NE(rv, -1, "setsockopt: ", strerror(errno));
+}
+
+void Socket::block(bool on) {
+  auto rv = fcntl(fd_, F_GETFL);
+  GLOO_ENFORCE_NE(rv, -1, "fcntl: ", strerror(errno));
+  if (!on) {
+    // Set O_NONBLOCK
+    rv |= O_NONBLOCK;
+  } else {
+    // Clear O_NONBLOCK
+    rv &= ~O_NONBLOCK;
+  }
+  rv = fcntl(fd_, F_SETFL, rv);
+  GLOO_ENFORCE_NE(rv, -1, "fcntl: ", strerror(errno));
+}
+
+void Socket::configureTimeout(int opt, std::chrono::milliseconds timeout) {
+  struct timeval tv = {
+      .tv_sec = timeout.count() / 1000,
+      .tv_usec = (timeout.count() % 1000) * 1000,
+  };
+  auto rv = setsockopt(fd_, SOL_SOCKET, opt, &tv, sizeof(tv));
+  GLOO_ENFORCE_NE(rv, -1, "setsockopt: ", strerror(errno));
+}
+
+void Socket::recvTimeout(std::chrono::milliseconds timeout) {
+  configureTimeout(SO_RCVTIMEO, std::move(timeout));
+}
+
+void Socket::sendTimeout(std::chrono::milliseconds timeout) {
+  configureTimeout(SO_SNDTIMEO, std::move(timeout));
+}
+
+void Socket::bind(const sockaddr_storage& ss) {
+  if (ss.ss_family == AF_INET) {
+    const struct sockaddr_in* sa = (const struct sockaddr_in*)&ss;
+    bind((const struct sockaddr*)sa, sizeof(*sa));
+    return;
+  }
+  if (ss.ss_family == AF_INET6) {
+    const struct sockaddr_in6* sa = (const struct sockaddr_in6*)&ss;
+    bind((const struct sockaddr*)sa, sizeof(*sa));
+    return;
+  }
+  GLOO_ENFORCE(false, "Unknown address family: ", ss.ss_family);
+}
+
+void Socket::bind(const struct sockaddr* addr, socklen_t addrlen) {
+  auto rv = ::bind(fd_, addr, addrlen);
+  GLOO_ENFORCE_NE(rv, -1, "bind: ", strerror(errno));
+}
+
+void Socket::listen(int backlog) {
+  auto rv = ::listen(fd_, backlog);
+  GLOO_ENFORCE_NE(rv, -1, "listen: ", strerror(errno));
+}
+
+std::shared_ptr<Socket> Socket::accept() {
+  struct sockaddr_storage addr;
+  socklen_t addrlen = sizeof(addr);
+  int rv = -1;
+  for (;;) {
+    rv = ::accept(fd_, (struct sockaddr*)&addr, &addrlen);
+    if (rv == -1) {
+      if (errno == EINTR) {
+        continue;
+      }
+      // Return empty shared_ptr to indicate failure.
+      // The caller can assume errno has been set.
+      return std::shared_ptr<Socket>();
+    }
+    break;
+  }
+  return std::make_shared<Socket>(rv);
+}
+
+void Socket::connect(const sockaddr_storage& ss) {
+  if (ss.ss_family == AF_INET) {
+    const struct sockaddr_in* sa = (const struct sockaddr_in*)&ss;
+    return connect((const struct sockaddr*)sa, sizeof(*sa));
+  }
+  if (ss.ss_family == AF_INET6) {
+    const struct sockaddr_in6* sa = (const struct sockaddr_in6*)&ss;
+    return connect((const struct sockaddr*)sa, sizeof(*sa));
+  }
+  GLOO_ENFORCE(false, "Unknown address family: ", ss.ss_family);
+}
+
+void Socket::connect(const struct sockaddr* addr, socklen_t addrlen) {
+  auto rv = ::connect(fd_, addr, addrlen);
+  if (rv == -1 && errno != EINPROGRESS) {
```

Comment:
Since failure is assert here, not sure if we want to handle EINTR? EAGAIN?

Reply:
Good point. Updating to handle EINTR. I don't want to deal with EAGAIN on `connect(2)`, because it means something is wrong (or at least poorly tuned):

> EAGAIN: Insufficient entries in the routing cache.

Reply:
SGTM